R/allelic_series_sumstats.R

```diff
@@ -0,0 +1,331 @@
+    out <- as.numeric(GetLambda(kernel))
+    return(out)
+  })
+
+  # Test parameters.
+  opt_params <- SkatOptimalParam(cov_z = cov_z, rhos = rhos)
+  results <- PerRhoResults(lambdas = lambdas, qstats = qstats, rhos = rhos)
+  pval <- OptimalPval(opt_params = opt_params, results = results)
+  return(pval)
+}
+
+
+# ------------------------------------------------------------------------------
+
+
+#' COding-variant Allelic Series Test from Summary Statistics
+#'
+#' Main function for performing the allelic series test from summary statistics.
+#' Performs both Burden and SKAT type tests, then combines the results to
+#' calculate an omnibus p-value. Note that not all tests included in
+#' \code{\link{COAST}} are available when working with summary statistics.
+#'
+#' @param anno (snps x 1) annotation vector with values in c(0, 1, 2).
+#' @param beta (snps x 1) vector of effect sizes for the coding genetic variants
+#'   within a gene.
+#' @param se (snps x 1) vector of standard errors for the effect sizes.
+#' @param check Run input checks? Default: TRUE.
+#' @param ld (snps x snps) matrix of correlations among the genetic variants.
+#'   Although ideally provided, an identity matrix is assumed if not.
+#' @param maf (snps x 1) vector of minor allele frequencies. Although ideally
+#'   provided, defaults to the zero vector.
+#' @param pval_weights (3 x 1) vector of relative weights for combining the
+#'   component tests when performing the omnibus test.
+#' @param weights (3 x 1) vector of annotation category weights.
+#' @return Numeric p-value.
+#' @examples 
+#' # Generate data.
+#' data <- DGP(n = 1e3)
+#' sumstats <- CalcSumstats(data = data)
+#' 
+#' # Run the Coding-variant Allelic Series Test from summary statistics.
+#' results <- COASTSS(
+#'   anno = sumstats$anno,
+#'   beta = sumstats$sumstats$beta, 
+#'   maf = sumstats$maf,
+#'   se = sumstats$sumstats$se,
+#'   ld = sumstats$ld
+#' )
+#' show(results)
+#' @export
+COASTSS <- function(
+    anno,
+    beta, 
+    se,
+    check = TRUE,
+    maf = NULL,
+    ld = NULL,
+    pval_weights = c(1, 1, 1),
+    weights = DEFAULT_WEIGHTS
+) {
+  
+  # Input checks.
+  if (check) {
+    CheckInputsSS(
+      anno = anno,
+      beta = beta,
+      se = se,
+      maf = maf,
+      ld = ld
+    )
+  }
+  
+  # Baseline model p-value.
+  p_base <- ASBTSS(
+    anno = anno,
+    beta = beta,
+    se = se,
+    check = FALSE,
+    ld = ld,
+    maf = maf,
+    method = "none",
+    weights = c(1, 1, 1)
+  )
+    
+  # Sum model p-value.
+  p_burden <- ASBTSS(
+    anno = anno,
+    beta = beta,
+    se = se,
+    check = FALSE,
+    ld = ld,
+    maf = maf,
+    method = "sum",
+    weights = weights
+  )
+  
+  # SKAT model p-value.
+  p_skat <- ASKATSS(
+    anno = anno,
+    beta = beta,
+    se = se,
+    check = FALSE,
+    ld = ld,
+    maf = maf,
+    weights = weights
+  )
+
+  # Omnibus p-value.
+  pvals <- c(p_base, p_burden, p_skat)
```

Comment:
not sure if we want to keep ACAT as one of the pvalues. For some reason, in real data, omnibus p-values with acat added consistently identifies more significant signals. (not sure if due to inflation though since we are using subG and id matrix for now.)

Reply:
I think we may want to exclude ACAT because it's essentially a single-variant test, and we have no way to ensure the provided sumstats are based on sufficient individuals for the per-variant p-values to be reliable.